packages/flutter/lib/src/services/asset_manifest.dart

```diff
@@ -0,0 +1,113 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+
+import 'asset_bundle.dart';
+import 'message_codecs.dart';
+
+const String _kAssetManifestFilename = 'AssetManifest.bin';
+
+/// Contains details about available assets and their variants.
+/// See [Asset variants](https://docs.flutter.dev/development/ui/assets-and-images#asset-variants)
+/// to learn about asset variants and how to declare them.
+abstract class AssetManifest {
+  /// Loads asset manifest data from an [AssetBundle] object and creates an
+  /// [AssetManifest] object from that data.
+  static Future<AssetManifest> loadFromAssetBundle(AssetBundle bundle) {
+    return bundle.loadStructuredBinaryData(_kAssetManifestFilename, _AssetManifestBin.fromStandardMessageCodecMessage);
+  }
+
+  /// Loads asset manifest data from the root bundle and creates an
+  /// [AssetManifest] from that data.
+  static Future<AssetManifest> loadFromRootBundle() {
+    return loadFromAssetBundle(rootBundle);
+  }
+
+  /// Lists the keys of all known assets, not including asset variants.
+  ///
+  /// The logical key maps to the path of an asset specified in the pubspec.yaml
+  /// file at build time.
+  ///
+  /// See [Specifying assets](https://docs.flutter.dev/development/ui/assets-and-images#specifying-assets)
+  /// and [Loading assets](https://docs.flutter.dev/development/ui/assets-and-images#loading-assets) for more
+  /// information.
+  List<String> listAssets();
+
+  /// Gets available variants of an asset.
+  List<AssetVariant> getAssetVariants(String key);
```

Comment:
Does this include a variant for the main asset?

Reply:
No it does not. I updated the doc string. However, this did make me realize something--this method should probably throw if `key` is not found within the manifest. Otherwise, it could be possible for users to mistakenly think that some asset has no variants when in reality they may be using some invalid hard-coded key. Do you agree?

Reply:
Thinking about this more, I want to get back to your original point. 

I think it also might be more intuitive from a user standpoint that an asset is not considered to be a variant of itself. However, it does make iterating over an asset and its variants more awkward. Also, it makes migration of code that consumes `AssetManifest.json` more error-prone since folks will have to remember to also operate on the main asset in addition to whatever loop they are replacing. I don't have a particular strong opinion on this at the very moment

Reply:
I agree with this, but I am having trouble figuring out an API design that I think will be intuitive to users.

Specifically, the [Assets and images](https://docs.flutter.dev/development/ui/assets-and-images) doc makes a distinction between a _main_ asset and its _variants_:

> Then both graphics/background.png and graphics/dark/background.png are included in your asset bundle. The former is considered the main asset, while the latter is considered a variant.

Therefore, I think having `getAssetVariants` include the main asset in its results conflicts with this distinction. I have a few ideas:
1. Update the doc to say that each main asset is also considered to be a variant of itself--but, again, I don't know if people would generally find that intuitive.
2. The doc string on `getAssetVariants` can mention that, specifically in the context of this method, the main asset is considered to be a variant of itself.
3. Rename the `AssetVariant` class to something more broad, like `AssetMetadata`, and add a method `AssetManifest.getAssetMetadata(String key)` for getting the metadata for an asset without having to find it indirectly through `getAssetVariants`.

I'm biased towards option 2 since its simple and I don't see the API being popular enough to worry so much over, but I am very much interested in any thoughts you have.

Reply:
I like a combination of 2/3. AssetMetadata sounds like a better name, and we can document what the return is. At any rate, there needs to be a way to get the data for the main asset if you decide to use it.